Iterators.md

```diff
@@ -0,0 +1,89 @@
+```
+Let’s break this down. We can start by looking at the two range functions, each of which constructs an iterator between the specified numbers. The `::` command links the two, instructing Coconut to proceed from the first iterator to the second. Following this, we use pipelines `(|>)` to tell Coconut to convert to a list and print the output:
+
+```
+Output([-10, -9, -8, -7, -6, -5, -4, -3, -2, -1])
+```
+
+This is all fine, but what is really special about Coconut’s iterator chaining abilities is that Coconut’s iterator chaining works lazily. This means that nothing gets evaluated until it is needed, which adds a lot of flexibility to Coconuts usage, especially when dealing with things like infinite series. Let’s take a look at the following function definition, which takes advantage of Coconut’s lazy chaining.
+
+```
+def N(n=0) = (n,) :: N(n+1)
+```
+
+At first glance this probably looks scary, or at the very least, like a recipe for an infinite loop. We see that the definition of the function `N` is that `n` is being concatenated with `n + 1`. However, this is not what will happen due to the lazy evaluation. Let’s use partial application to call this function on a range of values, then print the values iterated over to see what actually happens.
+
+```
+N()$[0:10] |> list |> print
+
+Output([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
+```
+
+Not so bad after all! If we were to try to define a function in this way in Python, it would be incredibly complicated, and it would require you to understand iterators and their functionality in Python on a very deep level. If you would like to get an idea for what this would look like, simply copy that one-line function definition above into a Coconut file and compile it to Python. It certainly isn’t pretty!
+
+To see a more complex implementation of chaining in coconut, look at Lazy Lists!
+
+Let’s now take a moment to look at some of the Python itertools functionality that’s built into Coconut and ready to use fairly quickly.
+
+## takewhile and dropwhile
+
+Coconut provides itertools.takewhile as a built-in part of the language under the name takewhile. It works the same way in both Python and Coconut. It takes a predicate or a condition as its first argument and then an iterable as its second argument and returns an iterator containing only those values that satisfied the condition specified. Let’s take a look at the following to begin with:
+
+```
+negatives = takewhile((x) -> x < 0, range(-10,10))
+```
+
+What does this say? Well we are defining an iterator called negatives. The iterable I have chosen to look through is the range of integers from -10 to 10. The predicate or condition is defined using Coconut’s simplified lambda syntax. It states that given an input `x`, we want to check if `x` is less than zero. The predicate and condition are fed to the takewhile function. Let’s convert negatives to a list and print to see what is returned when we iterate through negatives.
```

Comment:
Probably a lot easier and clearer to just write out what the function takes as input rather than describing it in text